### ISSUE TITLE
debug_handler: Check IntoResponseParts / IntoResponse for return tuple element types

### ISSUE THREAD

**Issue by jplatte** [tokio-rs/axum#2173]
In Jon Gjengset's [Decrusting the axum crate](https://youtu.be/Wnb_n5YktO8) video, he tried using `#[debug_handler]` to get a more useful error message for `(StatusCode, Json<_>, AppendHeaders<_>)` not implementing `IntoResponse` as part of a handler, but that didn't work. I think we can make it work.

**Comment by SpeedReach** [comment]
Is anyone working on this? I'd like to implement this if not.

**Comment by davidpdrsn** [comment]
I don’t think so. You’re welcome to submit a PR!

**Comment by SpeedReach** [comment]
Thanks for the opportunity! 
I improved the error message from
```
error[E0277]: the trait bound `(StatusCode, Json<&str>, AppendHeaders<[(HeaderName, &str); 1]>): IntoResponse` is not satisfied
 --> tests/debug_handler/fail/wrong_return_tuple_type.rs:5:23
  |
5 |   async fn handler() -> (
  |  _______________________^
6 | |     axum::http::StatusCode,
7 | |     axum::Json<&'static str>,
8 | |     axum::response::AppendHeaders<[( axum::http::HeaderName,&'static str); 1]>,
9 | | ) {
  | |_^ the trait `IntoResponse` is not implemented for `(StatusCode, Json<&str>, AppendHeaders<[(HeaderName, &str); 1]>)`
```
to
```
error[E0277]: the trait bound `Json<&'static str>: IntoResponseParts` is not satisfied
 --> tests/debug_handler/fail/wrong_return_tuple_type.rs:7:5
  |
7 |     axum::Json<&'static str>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoResponseParts` is not implemented for `Json<&'static str>`
```

It checks the
first element => StatusCode || Parts || IntoResponseParts
last element => IntoResponse
other elements => IntoResponseParts

I'll start to add tests if you guys think this is a good approach.

**Comment by jplatte** [comment]
That sounds good, but I think for the extractors we also have special checks for Json and other body extractors of known names where we then print a custom error message if they're not last. Those kinds of messages would be helpful here as well, I think.

**Comment by davidpdrsn** [comment]
Yes I agree that would be good but not essential and a PR without is also cool. We can always add that later.

**Comment by SpeedReach** [comment]
I was having the same thought when writing it, but I'm not sure what types should always be placed at last, I added a //todo 
where the logic should be written.

**Comment by SpeedReach** [comment]
Added special checks for known types that implements IntoResponse but not IntoResponseParts, but it only checks for Json and String now , the list can be extended.
Also opened a pr , let me know if there's change needed!

**Comment by SpeedReach** [comment]
Sorry this is my first time doing a pull request. Should I request a specific reviewer, or just leave it like that?

**Comment by davidpdrsn** [comment]
Nope you don’t need to do anything. I’ll take a look this week :)

**Comment by SpeedReach** [comment]
Got it, thanks!

**Comment by tahsin-s** [comment]
Was searching for issues to work on and found this; I believe this issue should be closed. 
[/pull/2201]

**Comment by mladedav** [comment]
You're right, thanks!
